apollo-router-core/src/query.rs

```diff
@@ -0,0 +1,469 @@
+use crate::prelude::graphql::*;
+use apollo_parser::ast;
+use derivative::Derivative;
+use std::collections::HashMap;
+
+#[derive(Debug, Derivative)]
+#[derivative(PartialEq, Hash, Eq)]
+pub struct Query {
+    string: String,
+    #[derivative(PartialEq = "ignore", Hash = "ignore")]
+    fragments: HashMap<String, Vec<Selection>>,
+    #[derivative(PartialEq = "ignore", Hash = "ignore")]
+    operations: Vec<Operation>,
+}
+
+impl Query {
+    /// Returns a reference to the underlying query string.
+    pub fn as_str(&self) -> &str {
+        self.string.as_str()
+    }
+
+    /// Re-format the response value to match this query.
+    ///
+    /// This will discard unrequested fields and re-order the output to match the order of the
+    /// query.
+    #[tracing::instrument(level = "trace")]
+    pub fn format_response(&self, response: &mut Response, operation_name: Option<&str>) {
+        let data = std::mem::take(&mut response.data);
+        match data {
+            Value::Object(init) => {
+                let output = self.operations.iter().fold(init, |mut input, operation| {
+                    if operation_name.is_none() || operation.name.as_deref() == operation_name {
+                        let mut output = Object::default();
+                        self.apply_selection_set(&operation.selection_set, &mut input, &mut output);
+                        output
+                    } else {
+                        input
+                    }
+                });
+                response.data = output.into();
+            }
+            _ => {
+                failfast_debug!("Invalid type for data in response.");
+            }
+        }
+    }
+
+    pub fn parse(
+        query: impl Into<String>,
+        schema: &Schema,
+    ) -> tokio::task::JoinHandle<Option<Self>> {
```

Comment:
I'm not sure about returning a join handle here (there's nothing that warrants access to the join handle at https://github.com/apollographql/router/pull/205/files#diff-830a30ed2f89feb3a5d8d08941aa96390c492fcceccc7d8ec4564ae800a712a6R74 ). Could we convert that to a more abstract future and hide the task spawning here?

I think putting the parsing in a blocking task should even be one level above, in the QueryCache code

Reply:
Fair point.

Fixed in 44fdb3cfefbc2d279181e28b1e0b07a1466c57f0

Reply:
nice, thanks